Use the SwiftIfConfig library for evaluating #if conditions

DougGregor opened the pull request:
Replace the C++ implementation of `#if` condition operator folding, validation, and evaluation with the corresponding facilities in the SwiftIfConfig library. Leave the C++ implementation in place for now to permit the compiler to continue building without swift-syntax.

There are a few differences in the diagnostics produced relative to the C++ implementation; the second commit shows the actual changes, all of which are benign or are improvements.

DougGregor committed:
Use the SwiftIfConfig library for evaluating #if conditions.

Replace the C++ implementation of #if condition operator folding,
validation, and evaluation with the corresponding facilities in the
SwiftIfConfig library.

Leave the C++ implementation in place for now to permit the compiler
to continue building without swift-syntax.

DougGregor commented:
Note that this builds on https://github.com/swiftlang/swift/pull/75014; only the last commit is new.

DougGregor committed:
Update expected diagnostics to match what SwiftIfConfig produces

For the most part, the differences between the diagnostics introduced
by the C++ implementation and the new SwiftIfConfig implementation are
cosmetic, so these are only wording changes.

The one major difference is that we've dropped the warnings about
potential typos in os/arch checks. For example, if one writes:

    #if os(bisionos)
    // ...
    #endif

The C++ implementation will produce a warning "unknown operating system
for build configuration 'os'" with a note asking "did you mean
'visionOS'"? These warnings rely on a static list of known operating
systems and architectures, which is somewhat unfortunate: the whole
point of these checks is that the Swift you're dealing with might not
have support for those operating systems/architectures, so while these
warnings can be helpful in a few cases, they also cause false
positives when porting. Therefore, I chose not to bring them forward.

DougGregor commented:
https://github.com/swiftlang/swift-syntax/pull/2783

@swift-ci please smoke test

ahoppen commented on lib/ASTGen/Sources/ASTGen/CompilerBuildConfiguration.swift:
```diff
@@ -150,6 +157,18 @@ struct CompilerBuildConfiguration: BuildConfiguration {
+  var description: String {
+    switch self {
+      case .unexpectedRuntimeCondition:
+        return "unexpected argument for the '_runtime' condition; expected '_Native' or '_ObjC'"
```
Isn’t `_multithreaded` also allowed, based on the check above?

DougGregor replied:
Heh, yes. I was originally following the actually-incorrect diagnostic text in the compiler itself to try to minimize the diffs in this PR. We can fix this now.

ahoppen commented on lib/ASTGen/Sources/ASTGen/CompilerBuildConfiguration.swift:
```diff
@@ -290,3 +309,53 @@ public func freeConfiguredRegions(
+) -> Int {
+  // Retrieve the #if condition that we're evaluating here.
+  // FIXME: Use 'ExportedSourceFile' when C++ parser is replaced.
+  let textBuffer = UnsafeBufferPointer<UInt8>(start: ifConditionText.data, count: ifConditionText.count)
```
Don’t we have an `ExportedSourceFile` already (the one that we also use for macro evaluation)? Or is this happening earlier where we haven’t created the `ExportedSourceFile` yet?

DougGregor replied:
I'm following the path @rintaro used for checking the definitions of macros, which avoids running the swift-syntax parser on the whole file (for performance reasons). This also avoids the somewhat brittle "go find this exact swift-syntax node at this source location" lookup we need to do with that approach, which can fail when the C++ parser and SwiftParser don't recover in exactly the same way.

ahoppen replied:
I see. I was still thinking we had the syntax tree for the entire file anyway and forgot that we no longer parse it completely.

DougGregor commented:
```
/home/build-user/swift/test/Parse/ifconfig_expr.swift:172:22: error: expected error not produced
#if canImport(A,) // expected-error {{unexpected ',' separator}}
                 ~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Whoops, bad rebase on my part now that [SE-0439](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0439-trailing-comma-lists.md) has landed and made this code well-formed.

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please test Windows

DougGregor commented:
Oh, I bet I have to fix my blatant layering violation for Windows .

DougGregor committed:
Requestify the #if computation using SwiftIfConfig

We're not caching this now, but it lets us dodge annoying layering
issues because ASTGen (where this is available) sits on top of the C++
parser, which needs to call it.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test Windows

DougGregor commented:
@swift-ci please test Windows

DougGregor commented:
@swift-ci please test Windows

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please test source compatibility

DougGregor commented:
https://github.com/swiftlang/swift-syntax/pull/2814

@swift-ci please test

DougGregor commented:
https://github.com/swiftlang/swift-syntax/pull/2814

@swift-ci please test source compatibility

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please test source compatibility
